_posts/2018-06-02-GraphQL-Stitching.markdown

```diff
@@ -0,0 +1,317 @@
+
+1.  Download the schema of each external API into metaphysics' source code
+1.  Have each schema trimmed to just the essentials that we need today
+1.  Merge in each schema incrementally
+1.  Stitch in any desired schema changes
+
+Let's dig, with some code into how we do each of these steps.
+
+#### Downloading Schemas
+
+We created a [pretty minimal script][dl-schema] which can be run periodically from a developer's computer.
+
+```js
+const destination = "src/data"
+
+const httpConvectionLink = createHttpLink({
+  fetch,
+  uri: urljoin("https://convection-staging.artsy.net/api", "graphql")
+})
+
+introspectSchema(httpConvectionLink).then(schema => {
+  fs.writeFileSync(path.join(destination, "convection.graphql"), printSchema(schema, { commentDescriptions: true }))
+})
+```
+
+The script uses an [apollo-http-link][] to grab our schema, and store it in our repo, see
+[`src/data/convection.graphql`][c-gql]. This means that when someone wants to update to a new version of the schema,
+it will go through code review and a normal testing-flow. The trade-off being that it will always be out of date a
+little bit, but you don't know how the schema will grow in the future.
+
+This file is the [GraphQL SDL][sdl] representations of the entire type system for that schema. This means we have a
+local copy of the schemas, so we can use it for tests for the next few steps.
+
+#### Schema Manipulation
+
+Each API writes for their own domain. This can be problematic when you use a `LineItem` in one API, which isn't
+generic enough to be a `LineItem` in a global API of all services. When thinking about this problem, we created a
+[guide for ourselves][guides-schema] on how to think about schema design at local and global level.
+
+We use a few of [the transform APIs][transform] available in graphql-tools to make the merges work. The first
+approach is to force a namespace by prefixing the merged Types with their domain.
+
+```js
+export const executableConvectionSchema = async () => {
+  const convectionLink = createConvectionLink()
+  const convectionTypeDefs = readFileSync("src/data/convection.graphql", "utf8")
+
+  // Setup the default Schema
+  const schema = await makeRemoteExecutableSchema({
+    schema: convectionTypeDefs,
+    link: convectionLink
+  })
+
+  // Remap the names of certain types from Convection to fit in the larger
+  // metaphysics ecosystem.
+  const remap = {
+    Submission: "ConsignmentSubmission",
+    Category: "ConsignmentSubmissionCategoryAggregation",
+    Asset: "ConsignmentSubmissionCategoryAsset",
+    State: "ConsignmentSubmissionStateAggregation",
+    SubmissionConnection: "ConsignmentSubmissionConnection"
+  }
+
+  // Return the new modified schema
+  return transformSchema(schema, [
+    new RenameTypes(name => {
+      const newName = remap[name] || name
+      return newName
+    })
+  ])
+}
+```
+
+Another example is to outright remove almost everything in the schema, and to only allow Types and fields which we
```

Comment:
Below there's a blacklist, not a list of what you want included, so this contradicts the example.